airsonic-main/src/main/webapp/WEB-INF/jsp/podcastChannel.jsp

```diff
@@ -1,45 +1,148 @@
+
+        // need to keep track if a request was sent because plaQueue may also send a request
+        var awaitingAppendPlaylistRequest = false;
+        function onAppendPlaylist() {
+            awaitingAppendPlaylistRequest = true;
+            // retrieve writable lists so we can open dialog to ask user which playlist to append to
+            top.StompClient.send("/app/playlists/writable", "");
+        }
+        function playlistSelectionCallback(playlists) {
+            if (!awaitingAppendPlaylistRequest) {
+                return;
+            }
+            awaitingAppendPlaylistRequest = false;
+            $("#dialog-select-playlist-list").empty();
+            for (var i = 0; i < playlists.length; i++) {
+                var playlist = playlists[i];
+                $("<p class='dense'><b><a href='#' onclick='appendPlaylist(" + playlist.id + ")'>" + escapeHtml(playlist.name)
+                    + "</a></b></p>").appendTo("#dialog-select-playlist-list");
+            }
+            $("#dialog-select-playlist").dialog("open");
+        }
+
+        function appendPlaylist(playlistId) {
+            $("#dialog-select-playlist").dialog("close");
+            var mediaFileIds = getSelectedEpisodesMediaId();
+            top.StompClient.send("/app/playlists/files/append", JSON.stringify({id: playlistId, modifierIds: mediaFileIds}));
+        }
+
+        function getMediaDirectory(mediaFileId) {
+            getMediaDirectories([mediaFileId]);
+        }
+
+        function getMediaDirectories(ids, paths) {
+            top.StompClient.send("/app/mediafile/directory/get", JSON.stringify({ids: ids, paths: paths}));
+        }
+        function getMediaDirectoryCallback(mediaDirObj) {
+            if (downloadMedia) {
+                var selectedMediaIds = getSelectedEpisodesMediaId();
+                var selectedFilesIndexesToDownload = mediaDirObj.files.reduce(function(accumulator, file, index) {
+                    if (selectedMediaIds.indexOf(file.id.toString()) > -1) {
+                        accumulator.push(index);
+                    }
+                    return accumulator;
+                }, []);
+                location.href = "download.view?id=" + ${model.channel.mediaFileId}  + "&" + selectedFilesIndexesToDownload.map(i => "i=" + i).join("&");
+                downloadMedia = false;
+            }
+        }
+
+        var downloadMedia = false;
+        // actionSelected() is invoked when the users selects from the "More actions..." combo box.
+        function actionSelected(id) {
+            if (id == "top") {
+                return;
+            } else if (id == "selectAll") {
+                selectAll(true);
+            } else if (id == "selectNone") {
+                selectAll(false);
+            } else if (id == "download") {
+                if (getSelectedEpisodesMediaId().length > 0) {
+                    downloadMedia = true;
+                    getMediaDirectory(${model.channel.mediaFileId});
```

Comment:
One, there is a security exposure here. Sure, in the UI its hidden behind a downloadrole, but in javascript, you're still giving folks the ability to at least ask the question.

Second, this has a race condition. The getSelectedEpisodesMediaId() might have changed between when the message was sent and when a reply was received (in getMediaDirectoryCallback())

Third, this seems like a very convoluted way to go about it. I think you're trying to retrieve the corresponding indices of the media file id in podcast channel directory? Wouldn't it be more prudent and straightforward to change the download controller to just accept multiple ids?

Reply:
Your absolutely right, "convoluted" sounds right :-) I had to use a lot of the websocket stuff, just to get the media id. I did all this for not needing to alter the download controller (noise reduction :-D), but actually it would have been better to just implement a change in the download controller ;-)

What do you think about if I would add a "p" parameter (instead of i) in the download controller which could point to the podcast episode id(s)?
In the download controller I would then need to get the mediaId's accordingly.

Reply:
I was thinking, you don't even need to add an additional parameter. Just make the id parameter accept lists. Then, since you already know mediafileids for all the selected episodes, pass those in directly!

Reply:
Had to stare at the code for I while to get it :-)
Got it... finally ;-)